File: advent_2021/day_01.py
"""
Advent of Code 2021 - Day 1
See: https://adventofcode.com/2021/day/1
"""

from collections import deque
from pathlib import Path

def part_a(depths):
    acc = 0
    prev = depths[0]
    for val in depths[1:]:
        if val > prev:
            acc += 1

        prev = val

    return acc

def part_b(depths):
    acc = 0
    buf = deque(depths[0:3], maxlen=3)
    prev_sum = sum(buf)
    for val in depths[3:]:
        buf.append(val)
        window_sum = sum(buf)
        if window_sum > prev_sum:
            acc += 1

        prev_sum = window_sum

    return acc

if __name__ == "__main__":
    p = Path(__file__).parent / "input" / 'day_01_a.txt'
    with open(p, "r") as f:
        values = [ int(x) for x in f.readlines() ]
        print(f"Answer for a is {part_a(values)}.")
        print(f"Answer for b is {part_b(values)}.")

File: advent_2021/src/bin/day_01.rs
use humantime::format_duration;
use regex::Regex;
use std::{
    fs::File,
    io::{BufRead, BufReader},
    time::Instant,
};

type AResult<T> = anyhow::Result<T>;

fn parse(lines: &[String]) -> AResult<Vec<u16>> {
    Ok(lines
        .iter()
        .map(|l| (*l).parse::<u16>())
        .map(Result::unwrap)
        .collect())
}

fn part_a(lines: &[String]) -> AResult<u16> {
    let mut it = parse(lines)?.into_iter();

    let mut acc = 0u16;
    let mut prev = it.next().unwrap();

    for next in it {
        acc += (next > prev) as u16;
        prev = next;
    }

    Ok(acc)
}

fn part_b(lines: &[String]) -> AResult<u16> {
    let parsed = parse(lines)?;
    let mut it = parsed.windows(3);

    let mut acc = 0u16;
    let mut prev = it.next().unwrap();

    for next in it {
        let psum: u16 = prev.iter().sum();
        let nsum: u16 = next.iter().sum();

        acc += (nsum > psum) as u16;
        prev = next;
    }

    Ok(acc)
}

fn main() -> AResult<()> {
    // Figure out which exercise we are so we can load the correct input file
    let pattern = Regex::new("[0-9]+$")?;
    let name = &std::env::args().next().expect("binary name not found.");
    let ex = pattern
        .find(name)
        .expect("binary name should contain a number")
        .as_str();
    println!("Running code for Day {}.", ex);

    // Load the appropriate input text
    let file = File::open(format!("./data/day_{ex}_a.txt"))?;
    let lines: Vec<String> = BufReader::new(file).lines()
        .map(Result::unwrap)
        .collect();

    // Run the solutions
    let start = Instant::now();
    println!("Part A result = {}", part_a(lines.as_slice())?);
    println!("Part B result = {}", part_b(lines.as_slice())?);
    let end = Instant::now();

    println!("Run took {}ms", format_duration(end - start));

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_INPUT: &str = "199
        200
        208
        210
        200
        207
        240
        269
        260
        263";

    #[test]
    fn test_a() -> AResult<()> {
        let lines: Vec<_> = TEST_INPUT.lines().map(|l| l.trim().to_string()).collect();
        assert_eq!(part_a(lines.as_slice())?, 7);
        Ok(())
    }

    #[test]
    fn test_b() -> AResult<()> {
        let lines: Vec<_> = TEST_INPUT.lines().map(|l| l.trim().to_string()).collect();
        assert_eq!(part_b(lines.as_slice())?, 5);
        Ok(())
    }
}
